### PR TITLE
SR with precomputed gradient pytrees

### PR DESCRIPTION
**Pull request by inailuig**
This so far only adds the necessary logic, adding the interfaces around it should be straightforward though.

- The gradients (oks) are stored as a pytree where the first dim of each leaf is k (of size n_samples)
- Reduces to the usual two matrix-vector multiplications if a flattened model and parameters  (and a flattened gradient matrix) are passed, all the tree maps should be optimized away by the jit complier
- ~Works with inhomogeneous parameters (e.g. a mix of complex and real) just like the onthefly code does~
- works with R->R, holomprphic C->C and R->C
-  non-homogenous parameters and non-holomorphic C->C requires converting to homogeneous real parameters
- is MPI enabled

### CODE REVIEW

File: netket/optimizer/sr/sr_treemv_logic.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# TODO cheapest way to calculate the gradients?
+
+
+@partial(jax.vmap, in_axes=(None, None, 0))
+def perex_grads(forward_fn, params, samples):
+    def f(p, x):
+        return forward_fn(p, jnp.expand_dims(x, 0))[0]
+
+    y, vjp_fun = jax.vjp(f, params, samples)
+    res, _ = vjp_fun(np.ones((), dtype=jnp.result_type(y)))
```
Comment (attila-i-szabo) [inline netket/optimizer/sr/sr_treemv_logic.py:37]: I don't think this handles R→C functions correctly, because only changes in the real part of the output are backpropagated, so you end up with the real part of the true gradients. This is explained (a bit abstrusely) in the Autodiff cookbook, here is a minimal demonstration:
```
def f(x):
    return x[0] + 1j*x[1]
y, vjp_f = jax.vjp(f, jnp.array([1.0,2.0]))
vjp_f(1.0+0.0j)

>>> (DeviceArray([1., 0.], dtype=float32),)
```
This is not an issue for the lazy solver, because you start with a `jvp` that works fine for R→C functions and apply `vjp` to a complex vector which picks out the right mix of real and imaginary components. Here I don't think you can get away with not splitting an R→C function into real and imaginary parts... But it's neat we can treat R→R and holomorphic cases in one go!

Comment (attila-i-szabo) [reply netket/optimizer/sr/sr_treemv_logic.py:37]: The alternative would be building the gradient matrix from `jvp` but that requires you do to do a full forward pass for every parameter in the net, which is prohibitively expensive. I think a `vmap`ped `grad` or `vjp` is the only cheap way of getting the gradient matrix

Comment (inailuig) [reply netket/optimizer/sr/sr_treemv_logic.py:38]: Thanks for catching this !

It was also an issue in the centered on-the-fly code in the part where the mean is precomputed. 
Now the only cases still failing are for inhomogeneous parameters.

The non-centered on-the-fly code (which arguably is the most elegant one and also faster than centered) is and was working correctly also with inhomogeneous parameters.

**Comment by attila-i-szabo** [comment]
Thanks for these fixes, I think it now works as it should!

I wanted to discuss with you whether it's better to autodetect the network type (as you do) or require the user to specify it (as I do). There are two use cases where I think the latter is better:

* The `R2R` code is the right choice for real-valued but signed wave functions too: log(ψ) has an imaginary part, but it is fixed to integer multiples of π, so it has no gradients and backpropagating it is just a waste of computer time; autodetection would use the R→C code, however.
* One has to make sure that C→C functions are holomorphic before using this code, simply because non-holomorphic functions don't have well-defined complex-valued gradients. (But they do have well-defined VJPs and JVPs, so the lazy code might still work for them, I didn't have time to do the maths.) Tbf all my code does for this is requiring a `holomorphic` switch, but an explicit contract might be better than silently assuming it...

What do you think? We should come up with a matching interface so that both kinds of S matrix could be generated by the same `SRJacobian` object depending on a switch. Maybe we could have a default differentiation-mode option that uses autodetect?

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from functools import partial
 from netket.stats import sum_inplace, subtract_mean
 from netket.utils import n_nodes
+import netket.jax as nkjax
```
Comment (gcarleo) [inline netket/optimizer/sr/sr_onthefly_logic.py:20]: sorry why do we import netket.jax here and not jax?

**Comment by PhilipVinc** [comment]
So this does not work for inhomogeneous parameters?

**Comment by attila-i-szabo** [comment]
Not as it's written now, no. In general, writing down an explicit S matrix for inhomogeneous parameters is not possible (basically, you both should and shouldn't take the real parts of the cross-terms), unless you split every complex parameter into its real and imaginary parts and treat them as real parameters. [`vjp` and `jvp` does this implicitly,](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#complex-numbers-and-differentiation) which is why the lazy code works for everything -- but even there, `to_dense` produces a meaningless S matrix for inhomogeneous parameters or non-holomorphic C→C functions. Here we have to store everything for calculating the correct S matrix, so we can't expect that these cases work, unless we break up all complex parameters into real ones.

**Comment by attila-i-szabo** [comment]
This is also why I would pass the differentiation mode as an explicit parameter, it's better if the user is aware of the limitations of the implementation.

**Comment by attila-i-szabo** [comment]
I thought more about how to make this implementation work with any kind of function. To allow full generality, we'd have to reduce every complex parameter to its real and imaginary part, which reduces everything else to the R→C differentiation mode. This is really necessary for non-holomorphic C→C functions, whose Jacobians can only be captured with 4 real numbers per parameter, there will be various redundancies (identically zero gradients, Cauchy-Riemann, etc.) for the other cases.

I think this complex to real reduction could be done reasonably painlessly with the PyTree interface. Basically, we would replace every leaf with a complex `dtype` with a `dict` with keys `"real"` and `"imag"` (or something more crazy to not clash with names users might want to use); the corresponding values are the real and imaginary parts of the original leaf, in the same layout as the original. This requires a `_cplx_to_dict` and a `_dict_to_cplx` routine, ideally written such that it can simply be decorated with `tree_map`. Then one would run the gradient calculation routine on `lambda W,σ: apply_fun(_dict_to_cplx(W), σ)`, and would pass `_cplx_to_dict(x)` as the RHS of `solve`.

Would this be an interesting direction? I think it's not something we desperately need straightaway (and I'd also keep the existing differentiation modes too because they have their advantages), but it might be something nice in the long term.

**Comment by inailuig** [comment]
> I think this complex to real reduction could be done reasonably painlessly with the PyTree interface. Basically, we would replace every leaf with a complex `dtype` with a `dict` with keys `"real"` and `"imag"` (or something more crazy to not clash with names users might want to use); the corresponding values are the real and imaginary parts of the original leaf, in the same layout as the original. This requires a `_cplx_to_dict` and a `_dict_to_cplx` routine, ideally written such that it can simply be decorated with `tree_map`. Then one would run the gradient calculation routine on `lambda W,σ: apply_fun(_dict_to_cplx(W), σ)`, and would pass `_cplx_to_dict(x)` as the RHS of `solve`.
> 
This is exactly what I already did for the test:
https://github.com/netket/netket/blob/750cb17c603d4716ccd4e27dfb790dc30fab00c0/Test/Optimizer/test_sr_onthefly_logic.py#L34
If needed we can just copypaste it from there.
The inverse transformation can be generated automatically with autodiff: https://github.com/netket/netket/blob/750cb17c603d4716ccd4e27dfb790dc30fab00c0/Test/Optimizer/test_sr_onthefly_logic.py#L43

**Comment by inailuig** [comment]
> This is also why I would pass the differentiation mode as an explicit parameter, it's better if the user is aware of the limitations of the implementation.

I think it would be nicer if the users could pass their model of choice and stuff would just *magically* work (like it seems to do with centered=False). However I can't think of any easy way of detecting holomorphic/nonholomorphic functions (apart from scanning the jaxpr for conj or symbolically checking the CR equations). Any Ideas? 

For now I tried to do some autodetection and converting to real (as suggested, I also can't think of cleaner way for nonhomogeneous and nonholomorphic) in the centereed onthefly code which has to precalculate the mean. 
@attila-i-szabo For the precomputed gradient pytrees we should be able to do the same.

I will split the fixes for O_mean in  onthefly into a separate PR so that the history is cleaner and we can merge that first.

**Comment by PhilipVinc** [comment]
I agree with clemens that having an automatic selection of the algorithm would be nice, though the possibility to override it would be welcome because we might have errors in our logic and/or people might want to experiment.

@inailuig @attila-i-szabo I would like us to proceed in little steps, also because there is the overarcing issue of #650 (redesigning the general SR interface) and i would like to avoid needless work.

My plan is the following:
 - This PR should contain only the logic for SRJacobian on pytrees. We shall merge it when ready @inailuig 
 - @attila-i-szabo 's PR #648 shall be merged when ready, aka when MPI is implemented and works with the original implementation. I'd like to merge that PR before adding support for pytrees and other things.

After, we finalize discussion on #650 and we implement the interface for managing the different solvers/SR matrix types.

**Comment by inailuig** [comment]
@PhilipVinc this is in principle ready, but we have to merge https://github.com/netket/netket/pull/661 first.

File: netket/jax/utils.py [hunk 3]
```diff
@@ -120,6 +143,84 @@ def maybe_promote_to_complex(*types):
         return main_typ
 
 
+def tree_conj(t):
+    r"""
+    conjugate all complex leaves
+    The real leaves are left untouched.
+
+    t: pytree
```
Comment (PhilipVinc) [inline netket/jax/utils.py:151]: can you put arguments in the style we use everywhere?
```python
from netket.utils.types import PyTree
def tree_conj(t: PyTree)
Args:
    t: blabla
````

File: netket/jax/utils.py [hunk 4]
```diff
@@ -120,6 +143,84 @@ def maybe_promote_to_complex(*types):
...
+        x,
+        target,
+    )
+
+
+def tree_axpy(a, x, y):
+    r"""
+    compute a * x + y
+
+    a: scalar
```
Comment (PhilipVinc) [inline netket/jax/utils.py:194]: idem

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 5]
```diff
@@ -91,16 +40,32 @@ def O_vjp(x, params, v, forward_fn):
     return jax.tree_map(sum_inplace, res)  # allreduce w/ MPI.SUM
 
 
-def O_mean(samples, params, forward_fn):
+def O_mean(samples, params, forward_fn, holomorphic=True):
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_onthefly_logic.py:43]: can you put arguments in the order we use normally? `forward_fun, params, samples` (a bit everywhere

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 6]
```diff
@@ -91,16 +40,32 @@ def O_vjp(x, params, v, forward_fn):
...
+    w = jnp.ones(samples.shape[0], dtype=dtype) * (1.0 / (samples.shape[0] * n_nodes))
+
+    homogeneous = nkjax.tree_ishomogeneous(params)
+    real_params = not nkjax.tree_leaf_iscomplex(params)
+    real_out = not nkjax.is_complex(jax.eval_shape(forward_fn, params, samples))
+
+    if homogeneous and (real_params or holomorphic):
+        if real_params and not real_out:
+            # R->C
+            return nkjax.vjp(forward_fn, params, samples)[1](w)[0]
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_onthefly_logic.py:60]: in O_vjp you run mpi.allreduce but here you don't?

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 7]
```diff
@@ -91,16 +40,32 @@ def O_vjp(x, params, v, forward_fn):
...
-    v = jnp.ones(samples.shape[0], dtype=dtype) * (1.0 / (samples.shape[0] * n_nodes))
-    return O_vjp(samples, params, v, forward_fn)
+    w = jnp.ones(samples.shape[0], dtype=dtype) * (1.0 / (samples.shape[0] * n_nodes))
+
+    homogeneous = nkjax.tree_ishomogeneous(params)
+    real_params = not nkjax.tree_leaf_iscomplex(params)
+    real_out = not nkjax.is_complex(jax.eval_shape(forward_fn, params, samples))
+
+    if homogeneous and (real_params or holomorphic):
+        if real_params and not real_out:
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_onthefly_logic.py:58]: i don't understand what changes between nkjax.vjp and O_vjp

Comment (inailuig) [reply netket/optimizer/sr/sr_onthefly_logic.py:58]: its calling https://github.com/netket/netket/blob/8a0c4658879a01c3738c393fdb789f9d31d931d7/netket/jax/_vjp.py#L96

**Comment by PhilipVinc** [comment]
Ops i just realised i reviewed the wrong one

File: netket/optimizer/sr/sr_onthefly_logic.py (hunk 7 above)
Comment (PhilipVinc) [reply netket/optimizer/sr/sr_onthefly_logic.py:58]: and what does O_vjp does that nkjax.vjp does not do?

Comment (PhilipVinc) [reply netket/optimizer/sr/sr_onthefly_logic.py:58]: minus the allreduce

**Comment by PhilipVinc** [comment]
Can you rebase it on master?

File: netket/optimizer/sr/sr_treemv_logic.py [hunk 8]
```diff
@@ -41,12 +41,12 @@ def f(p, x):
 
 @partial(jax.vmap, in_axes=(None, None, 0))
 def perex_grads_rc(forward_fn, params, samples):
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_treemv_logic.py:43]: Is this called perex just because we used to call something perex in the past?
I never got that name...
If you want to pick a better name the honour is yours.

Comment (inailuig) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: Its called perex_grads because it was used in the old netket, I think it comes from one of the jax tutorials...

do you have a suggestion?

Comment (PhilipVinc) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: Yes i know. I personally never liked it. I don't think it tells what this really is
But if you think it is, that's fine by me.
This is like a vmapped grad, no?

Comment (attila-i-szabo) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: The equivalent function in the matrix-based code is just called `gradients`. I think that's descriptive enough given the file context

Comment (inailuig) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: perex -> per-example

But what it does is calculate the jacobain in a more efficient way

**Review by PhilipVinc — APPROVED** [review]

File: netket/optimizer/sr/sr_treemv_logic.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def sub_mean(oks):
+    return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
+
+
+def prepare_doks(forward_fn, params, samples):
+    oks = perex_grads(forward_fn, params, samples)
+    n_samp = samples.shape[0] * n_nodes  # MPI
+    # TODO where to divide by n_samp?
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_treemv_logic.py:69]: ?

File: netket/optimizer/sr/sr_treemv_logic.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    else:
+        return perex_grads_rr_cc(forward_fn, params, samples)
+    # TODO inhomogeneous R&C -> C
+
+
+def sub_mean(oks):
+    return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
+
+
+def prepare_doks(forward_fn, params, samples):
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_treemv_logic.py:66]: can you add a docstring saying what this does?

File: netket/optimizer/sr/sr_treemv_logic.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    res = jax.tree_map(partial(jnp.tensordot, w, axes=1), oks)
+    return jax.tree_map(sum_inplace, res)  # MPI
+
+
+def _mat_vec(v, oks):
+    res = tree_conj(vjp(oks, jvp(oks, v).conjugate()))
+    return tree_cast(res, v)
+
+
+def mat_vec(v, oks, diag_shift):
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_treemv_logic.py:95]: and to this one, saying what the expected inputs are and what it comptues

File: netket/optimizer/sr/sr_treemv_logic.py (hunk 8 above)
Comment (inailuig) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: I am also in favour of renaming, it took me a while to find what perex means...

Comment (PhilipVinc) [reply netket/optimizer/sr/sr_treemv_logic.py:43]: vmap_grad with a note that it's a more efficient implementation?

File: netket/optimizer/sr/sr_treemv_logic.py (hunk 9 above)
Comment (inailuig) [reply netket/optimizer/sr/sr_treemv_logic.py:69]: I wasn't sure wether it would be better to divide the oks by the sqrt like @attila-i-szabo did, or divide the intermediate or final result of the marix-vector product (n_samp might be large)

Comment (inailuig) [inline netket/optimizer/sr/sr_treemv_logic.py:69]:
```suggestion
```

Comment (attila-i-szabo) [reply netket/optimizer/sr/sr_treemv_logic.py:69]: As far as floating-point arithmetic goes, these are all equally good as long as n_samp isn't like 10^100... Maybe my solution is the least transparent of all, I wasn't sure which one to go for, either...

Comment (attila-i-szabo) [reply netket/optimizer/sr/sr_treemv_logic.py:69]: Where do you divide then?

**Comment by PhilipVinc** [comment]
Looks good to me now. Do you still have something you want to do or can I merge?

**Comment by inailuig** [comment]
@PhilipVinc feel free to merge
